---
output: 
  pdf_document:
    highlight: haddock
    df_print: kable
    extra_dependencies: ["flafter"]
urlcolor: blue
---

# Summary statistics and quality of the *Dictyostelium discoideum* reference genome assembly

#### Jason Moggridge

#### 29/01/2021

------

```{r setup, echo=FALSE, message=FALSE}
# Set so that long lines in R will be wrapped:
# knit
knitr::opts_chunk$set(
	message = FALSE,
	warning = FALSE,
	tidy = TRUE,
	tidy.opts = list(width.cutoff = 80)
)
library(tidyverse, quietly = T)
library(patchwork)
library(rcartocolor)
```

## Introduction

In this work, I examine the reference genome of the social amoeba *Dictyostelium discoideum* (Eichinger *et al*., 2005). *D. discoideum* is a interesting model organism for many areas of research in molecular biology, particularly because of their ability to cooperatively form a multicellular super-organism for the purpose of sporulation under starvation conditions (Williams, 2010). The *de-novo* genome assembly was created by whole chromosome shotgun sequencing using the Sanger method. For this, chromosomes were separated using PFGE. Isolated chromosomes were fragmented and then cloned using plasmids and yeast artificial chromosomes (YACs). As this was among the first protozoan genome projects, there was no draft genome to guide contig assembly. As such, data from HAPPY mapping, previously mapped genes, and YACs were used in creating the assembly. HAPPY mapping is akin to linkage mapping but uses random DNA fragmentation and PCR to get distance information for markers (sequence-tagged sites in this case) instead of cloning (Dear and Cook, 1993). This information was used to assign reads from each library to chromosome-specific bins through `BLAST` or `Atlas Overlapper`. Binned reads were then joined into contigs by either `GAP4` or `PHRED/PHRAM/CONSED` software. Then read-pair data and `BLAST` searches of all sequence data were used to extend contigs and create scaffolds. Gap were closed by various strategies, depending on whether sequencing was challenged by repetitive or A+T rich regions. The authors presented an analysis of the genome's nucleotide composition, repeats, transposons, tRNAs, telomeres, and centromere regions. In addition, gene prediction and translation was performed to create a phylogeny to place *D. discoideum* among other eukaryotic phyla based on the predicted proteome.

**Links**
- [(Browse genome record on NCBI)](https://www.ncbi.nlm.nih.gov/assembly/GCF_000004695.1/)
- [(Zipped FASTA file through ftp)](https://ftp.ncbi.nih.gov/genomes/refseq/protozoa/Dictyostelium_discoideum/all_assembly_versions/GCF_000004695.1_dicty_2.7/GCF_000004695.1_dicty_2.7_genomic.fna.gz)


----

## Analysis of the *D. discoideum* genome
  
*Lines starting with '$' are unix commands and '#' are comments, otherwise lines are R code*

<!-- I obtained the *Dictyostelium discoideum* reference genome from NCBI Refseq as follows: -->

<!-- ``` -->
<!-- $ rsync rsync://ftp.ncbi.nih.gov/genomes/refseq/protozoa/Dictyostelium_discoideum/ -->
<!--   all_assembly_versions/GCF_000004695.1_dicty_2.7/ -->
<!--   GCF_000004695.1_dicty_2.7_genomic.fna.gz  ./ -->
<!-- $ gunzip GCF_000004695.1_dicty_2.7_genomic.fna.gz -->
<!-- $ mv GCF_000004695.1_dicty_2.7_genomic.fna ./dicty_genome.fna -->
<!-- ``` -->

### The *D. discoideum* assembly has a size of 34.2 Mbp.
  
```
# remove lines with headers (>), strip newline characters (\n), count characters
$ grep -v '>' dicty_genome.fna | tr -d '\n' | wc -c
```

### There are 41 scaffolds in the assembly.

These include 6 genomic chromosomes, the mitochondrial chromosome and a plasmid chromosome, as well as 33 unplaced scaffolds.

```
# count scaffold headings
$ grep '>' dicty_genome.fna | wc -l
```

```{r message=FALSE, warning=FALSE, echo = FALSE}
seqlen <- read_delim('data/seqlens.tsv', '\t')
```
### The mean scaffold length is `r round(mean(seqlen$Length)/1000, 1)` kbp.   

Scaffold lengths are plotted in fig. 1.
```
# create a table with headings: Scaffold, Length
$ touch seqlens.tsv
$ echo -e 'Scaffold\tLength' > seqlens.tsv
# loop over headings, get length of following sequence for each
$ awk '/^>/ {if (seqlen){print seqlen}; printf $0"\t";seqlen=0;next; } 
{seqlen += length($0)}END{print seqlen}' dicty_genome.fna >> seqlens.tsv
```


### N50 & L50

I was able to recreate the N50  and L50 statistics reported on NCBI (5,450,249 bp and 3) as follows:

```{r}
N50 <- function(scaffold_lengths){
  lengths <- sort(scaffold_lengths, decreasing = T)
  total <- sum(lengths); cumsum <- 0
  for (i in seq_along(lengths)){
    cumsum <- cumsum + lengths[i]
    if (cumsum >= total/2) return(lengths[i])
  }
}
seqlen <- read_delim('data/seqlens.tsv', '\t')
dicty.N50 <- N50(seqlen$Length)
dicty.L50 <- which(sort(seqlen$Length, decreasing = T) == dicty.N50)
```


### Ns are rare in the assembly (0.07 %).

```
# N-content; total bp
$ grep -v '^>' dicty_genome.fna | tr -cd 'Nn' | wc -c
$ grep -v '>' dicty_genome.fna | tr -d '\n' | wc -c
```
```{r}
N_percent <- 23142 * 100 / 34204973
```
<!-- Only a very small portion (`r round(N_percent, 3)` %) of the assembly are ambiguous bases (N). -->


### The ratio of repetitive to unique sequence is `r round(18093064/16088767, 2)`.

```
# repetitive; unique 
$ grep -v '>' dicty_genome.fna | tr -d '\n' | tr -cd acgt | wc -c
$ grep -v '>' dicty_genome.fna | tr -d '\n' | tr -cd ACGT | wc -c
# 18093064, 16088767
```


### Table 1: Less than 1% of the assembly is in unplaced scaffolds

```{r message=FALSE, warning=FALSE, tidy=FALSE}
seqlen <- seqlen %>% 
  mutate(Type = case_when(
           str_detect(Scaffold, 'chrUn_') ~ 'Unplaced',
           str_detect(Scaffold, 'plasmid') ~ 'Plasmid',
           str_detect(Scaffold, 'mitochond') ~ 'Mitochondrial',
           TRUE ~ 'Nuclear chr.'))
seqlen %>% 
  group_by(Type) %>% 
  summarize(`Percent of Genome` = round(sum(Length)/sum(seqlen$Length)*100, 2))
```

**The *D. discoideum* genome is very A+T-rich**. All bases were counted and these are presented in fig. 1.
```
$ touch base_count.csv; echo "Base, Count" > base_count.csv
$ for base in A C G T a c g t
$ do
$   count=$(cat dicty_genome.fna | grep -v '>' | tr -cd $base | wc -c) 
$   echo "$base, $count" >> base_count.csv
$ done
```

----

```{r fig.height=3.85, fig.width=5.5, message=FALSE, warning=FALSE, echo=FALSE}

ggplot2::theme_set(theme_classic())
A <- seqlen %>% 
  arrange(desc(Length)) %>% 
  mutate(Rank = row_number()) %>% 
  ggplot() + 
  geom_point(aes(x = Rank, y = Length, colour = Type)) +
  scale_y_log10() +
  scale_color_carto_d() +
  labs(x='Scaffold', y = 'Scaffold length (bp)') +
  theme(legend.position = 'right', 
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 8))

base_comp <- read.csv('data/base_count.csv')
B <- base_comp %>%  
  mutate(Group = c(rep('Unique', 4), rep('Repetitive', 4)),
         Base = toupper(Base)) %>%
  ggplot() +
  geom_col(aes(x = Group, y = Count, fill = Base)) +
  labs(x = 'Type of sequence', y = 'Count')
C <- base_comp %>% 
  mutate(Base = toupper(Base), Proportion = Count/sum(Count)) %>% 
  ggplot() +
  geom_col(aes(x = 'Full genome', y = Proportion, fill = Base)) +
  labs(x = '')
BC <- B+C + plot_layout(widths = c(3.5,2), guides= 'collect')
# create figure
plt.title <- expression(paste(
  'Figure 1: ', italic("D. discoideum"), 
  " scaffold lengths and nucleotide composition"))

A/(BC) + 
  plot_annotation(
    subtitle = plt.title, 
    caption = expression(
      italic("*Code omitted for brevity")),
    tag_levels = 'A'
    )
```



## Discussion

The *D. discoideum* reference genome is relatively small (34.2 Mbp) and the assembly is of remarkably high quality and completeness. The assembly has nearly complete scaffolds for the six nuclear chromosomes, with very few unplaced scaffolds (33; fig. 1). The percentage of the assembly in these unplaced scaffolds is only 0.56 %. Correspondingly, the assembly has a very large N50 (5.45 Mbp) and a very small L50 (3 scaffolds). The sequence data is of high quality, with only 0.07 % ambiguous bases (N). The genome contains a relatively even split between repetitive sequence and unique sequence (ratio = 1.12). 

The  quality of the genome is surprisingly good, given that the effort was reliant on cloning and Sanger sequencing technology of the time, and dealt with complex repetitive regions that are difficult to assemble, as well as A+T-rich tracts (fig. 2) that can be difficult to clone and sequence (Eichinger *et al.*, 2011). To create this near-perfect *de-novo* genome assembly, the authors applied data from contiguity methods that were commonly used at the time: genetic mapping (HAPPY maps), physical mapping (with YACs, *etc*.), and paired reads. Overall, this reference genome should prove to be an invaluable resource for studying this intriguing model amoeba.

----

## References

- Dear PH, and PR Cook. 1993. Happy mapping: linkage mapping using a physical analogue of meiosis. *Nucleic Acids Res*. 11; 21(1): 13–20.

- Eichinger L, *et al.*. 2005. The genome of the social amoeba Dictyostelium discoideum. *Nature*. May 5; 435(7038): 43–57.

- Williams JG. 2010. Dictyostelium finds new roles to model. *Genetics*, 185(3):717–726.

